### PR TITLE
MLDB-2141 stop following redirects forever

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: testing/http_streambuf_test.cc [hunk 1]
```diff
@@ -32,68 +32,83 @@ using namespace MLDB;
...
-        socklen_t addrLen = sizeof(addr);
-        if (getsockname(sockfd, (struct sockaddr *)&addr, &addrLen) == -1) {
-            throw Exception("Failed to getsockname");
-        }
-        freePort = addr.sin_port;
-        close(sockfd);
-    }
-    string baseUrl = "http://localhost:" + to_string(freePort) + "/";
-
+struct TestServer {
```
Comment (FinchPowers) [inline testing/http_streambuf_test.cc]: I extracted the TestServer so that it runs for the whole test.

File: testing/http_streambuf_test.cc [hunk 2]
```diff
@@ -32,68 +32,83 @@ using namespace MLDB;
...
 
+    ~TestServer() {
+        // Unsafe, those may throw, but it's "only" a test...
+        runner.kill();
+        runner.waitTermination();
+    }
+
+} testServer;
+
+BOOST_AUTO_TEST_CASE( test_error_codes )
```
Comment (FinchPowers) [inline testing/http_streambuf_test.cc]: Same test as before, with the TestServer extracted.

File: testing/http_streambuf_test.cc [hunk 3]
```diff
@@ -113,7 +128,34 @@ BOOST_AUTO_TEST_CASE( test_error_codes )
         string expect = "HTTP code " + to_string(code) + " reading http://";
         BOOST_REQUIRE(strCerr.str().find(expect) != string::npos);
     }
+}
+
+BOOST_AUTO_TEST_CASE( test_infinite_redirect )
```
Comment (FinchPowers) [inline testing/http_streambuf_test.cc]: This is the new test.

**Comment by guyd** [comment]
Nice catch +1

File: http/http_rest_proxy.cc [hunk 4]
```diff
@@ -330,6 +330,8 @@ perform(const std::string & verb,
 
         myRequest.add_option(CURLOPT_URL, uri);
 
+        myRequest.add_option(CURLOPT_MAXREDIRS, 20);
```
Comment (jraby) [inline http/http_rest_proxy.cc]: I would suggest to keep this guy close to `CURLOPT_FOLLOWLOCATION` below, guarded by `if followRedirect`

**Comment by mathieumb** [comment]
Looks good. Thanks for the explanations in the PR. +1

**Pull request merged**